### PR TITLE
Add the policy class to the `NotAuthorizedError` message

### PR DESCRIPTION
**Pull request by smitssjors**
This PR fixes #681 and #647.

### CODE REVIEW

**Comment by Burgestrand** [comment]
Hi! Thanks for updating the old PR!

This does _not_ fix #647, but it does fix #681.

I'm not sure why the pull request templates we have didn't get proposed, but please have a look at it: https://github.com/varvet/pundit/blob/main/.github/PULL_REQUEST_TEMPLATE/pull_request_template.md - this PR is missing a changelog entry 🙂

**Comment by smitssjors** [comment]
I updated the PR to now actually take care of #647 and I updated the `CHANGELOG.md`

I did not get the template when I made the PR via the button in my fork. I did however read `CONTRIBUTING.md` so I added updating the CHANGELOG as a step to hopefully prevent these mistakes in the future.

**Comment by Burgestrand** [comment]
The jruby failures are unrelated to this change. Something is up with the more recent jruby version on our GitHub CI pipeline.

**Comment by Burgestrand** [comment]
Either way I think this should probably be good! I'll most likely merge this soon. Travelling to Baltic Ruby today though 😊

File: lib/pundit.rb [hunk 1]
```diff
@@ -36,7 +36,9 @@ def initialize(options = {})
         @record = options[:record]
         @policy = options[:policy]
 
-        message = options.fetch(:message) { "not allowed to #{query} this #{record.class}" }
+        record_name = record.is_a?(Class) ? record.name : "this #{record.class}"
+
+        message = options.fetch(:message) { "not allowed to #{policy.class}##{query} #{record_name}" }
```
Comment (Burgestrand) [inline lib/pundit.rb:41]: Hi! I've thought about this some more and I've got a change that's worth making.

The reasons for the change are:
1. `record.to_s` works better for anonymous classes, and it's what we use in PolicyFinder so it's somewhat battle-tested.
2. If there's an explicit message we should not calculate a record name.

```suggestion
        message = options.fetch(:message) do
          record_name = record.is_a?(Class) ? record.to_s : "this #{record.class}"
          "not allowed to #{policy.class}##{query} #{record_name}"
        end
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+- Improve the `NotAuthorizedError` message to include the policy class.
```
Comment (Burgestrand) [inline CHANGELOG.md:5]: Please add the issue number here :)

```suggestion
- Improve the `NotAuthorizedError` message to include the policy class. (#812)
```

**Review by Burgestrand — CHANGES_REQUESTED** [review]

**Review by Burgestrand — APPROVED** [review]

**Comment by Burgestrand** [comment]
Thank you!

**Pull request merged**